### ISSUE TITLE
Cuda in eko

### ISSUE THREAD

**Issue by scarlehoff** [N3PDF/eko#56]
As we discussed the other day (/cc @felixhekhorn @AleCandido) I've tried to run `eko` in a GPU. It doesn't work out of the box right now, but the only thing stopping the `integrand` to be a cuda kernel is the `cern_polygamma` function and the scipy zeta function. The rest seems to compile ok (besides some `numpy` functions that have to be massaged a bit, but nothing too dramatic).

If we wanted to run `eko` in GPU we would just have to create a kernel for some of these, but nothing too complicated.

Ps: then there are some problems that should not be, such as `math.gamma` or `np.exp(complex)` which in theory is supported but worst-case-scenario can also be kernel'd in or changed to things as `cmath.exp` which work fine.

**Comment by scarrazza** [comment]
Thanks for checking, but when you say that running on GPU is possible you are talking about execution, not strategy to get performance improvements, right?

**Comment by scarlehoff** [comment]
Just execution, but the vectorization should come from doing things like the integrand I guess.

**Comment by AleCandido** [comment]
@scarlehoff probably you already had a look to the `cern_polygamma` function, but I also looked and the only two things that are making it different from plain fortran-like code are:
1. raising an error (but there are others `numba` compiled functions that are raising errors, and you said there was no issue with them)
2. using `nb.List`, but it's a numba thing, so it should be fine

Than of course `cern_polygamma` it's using some `numpy` functions, and I know nothing about `scipy.special.zeta`, but do you have any guess about what could the error be?

Also we released today a new version, and now we typed everything, so `numba` it's compiling ahead of time and it's also changing behaviour at runtime, taking much less (compared to the previous time w/o compilation), so maybe it would be interesting to try with this one, if it doesn't cost you anything (otherwise don't mind). Do you tried with `master` or do you already used `develop`?

**Comment by scarlehoff** [comment]
> @scarlehoff probably you already had a look to the `cern_polygamma` function, but I also looked and the only two things that are making it different from plain fortran-like code are

Being Fortran code is one of the issues :P 

Just to clarify, when I say stopping I mean from working ootb. It can be massaged out (even the nb.list -being numba doesn't make it cudafriendly though-, it's just allocating a constant array), I just skipped it.

> scipy.special.zeta, but do you have any guess about what could the error be?

Just the fact that numba wouldn't know what is the corresponding cuda kernel. The only supported functions are: https://numba.readthedocs.io/en/stable/cuda/cudapysupported.html#built-in-types (and I have found a few that are not 100% supported but they are working on them).

> Do you tried with master or do you already used develop?

Master.

**Comment by felixhekhorn** [comment]
@scarlehoff since both the `eko` code and `numba` evolved quite a lot - maybe you can retry this at some point? However, of course, this is not urgent

**Comment by scarlehoff** [comment]
What part would you want to see offloaded to the GPU? My guess would be the integrand (as it is what I naively would think benefits enormously of being parallelized) but is that the/a bottleneck when actually running eko?

**Comment by felixhekhorn** [comment]
Indeed: there are two integrations that take most of the time: the actual evolution kernel and the matching conditions (and soon there will be yet a third time: the splitting functions for scheme B - right, @giacomomagni ? )

**Comment by giacomomagni** [comment]
Yes @felixhekhorn you are right. The third type of integration has to be performed just once independently from the evolution path.
